src/main/kotlin/com/jetbrains/snakecharm/codeInsight/SnakemakeSectionsInShellReferenceContributor.kt

```diff
@@ -86,17 +91,38 @@ class SnakemakeSectionsInShellReferenceContributor : PsiReferenceContributor() {
     private fun addSectionReferences(
             element: PyStringLiteralExpression,
             sectionMatcher: Matcher,
-            sectionReferences: MutableList<PsiReference>
+            sectionReferences: MutableList<PsiReference>,
+            toplevelVariables: List<PyTargetExpression>
     ) {
         while (sectionMatcher.find()) {
             val sectionName = sectionMatcher.group(1)
+                    ?.replace(CompletionInitializationContext.DUMMY_IDENTIFIER_TRIMMED, "")
             val argumentName = sectionMatcher.group(3)
-            if (sectionName.isEmpty() || ALLOWED_IN_SHELL_WITHOUT_KEYWORDS.any { it.startsWith(sectionName) }) {
+
+            if (sectionName == null) {
+                continue
+            }
+
+            val variable = toplevelVariables.find { it.name == sectionName }
+            if (variable != null) {
+                sectionReferences.add(SmkVariableReference(
+                        element,
+                        TextRange(sectionMatcher.start(1), sectionMatcher.start(1) + sectionName.length),
+                        variable
+                ))
+            } else if (sectionName.isEmpty() || ALLOWED_IN_SHELL_WITHOUT_KEYWORDS.any { it.startsWith(sectionName) }) {
                 sectionReferences.add(SmkSectionReference(
                         element,
-                        TextRange(sectionMatcher.start(1), sectionMatcher.end(1)),
+                        TextRange(sectionMatcher.start(1), sectionMatcher.start(1) + sectionName.length),
                         if (sectionName.isEmpty()) null else sectionName
                 ))
+            } else {
+                // unresolved reference for variable
+                sectionReferences.add(SmkVariableReference(
```

Comment:
You cannot grant here the correctness of the solution. Please do not show 'red errors' when you cannot be 100% sure. Use warnings/weak warnings. Maybe `PyInspectionExtension` could be useful for such cases.
I suggest to extract commits related to "top-level" vars resolve to a separate branch and focus here on #3 related features. Because top-level vars resolve is a separate big feature and better not to mix it with #3.